File: typescript/tests/integration.mjs
import { PipeStorageClient } from "../dist/index.js";

function requiredEnv(name) {
  const value = process.env[name];
  if (!value) {
    throw new Error(`Missing required env var: ${name}`);
  }
  return value;
}

async function main() {
  requiredEnv("PIPE_API_KEY");

  const fileName = `agent/integration-${Date.now()}.json`;
  const marker = `marker-${Date.now()}`;

  const client = new PipeStorageClient({
    apiKey: process.env.PIPE_API_KEY,
    account: process.env.PIPE_ACCOUNT,
    baseUrl:
      process.env.PIPE_BASE_URL ??
      process.env.PIPE_API_BASE_URL ??
      "https://us-west-01-firestarter.pipenetwork.com",
    timeoutMs: Number(process.env.PIPE_TEST_TIMEOUT_MS ?? 180000),
    pollIntervalMs: Number(process.env.PIPE_TEST_POLL_MS ?? 1000),
  });

  console.log("1) store(wait=false)", { fileName });
  const stored = await client.store({ marker, kind: "integration" }, { fileName, wait: false });
  if (!stored.operationId) {
    throw new Error("store did not return operationId");
  }

  console.log("2) checkStatus", { operationId: stored.operationId });
  const status1 = await client.checkStatus({ operationId: stored.operationId });
  console.log("   status=", status1.status);

  console.log("3) waitForOperation");
  const completed = await client.waitForOperation(stored.operationId, {
    timeoutMs: Number(process.env.PIPE_TEST_TIMEOUT_MS ?? 180000),
  });
  if (completed.status !== "completed") {
    throw new Error(`unexpected final status: ${completed.status}`);
  }

  console.log("4) pin");
  const pinned = await client.pin({ operationId: stored.operationId });
  if (!pinned.url) {
    throw new Error("pin did not return deterministic URL");
  }

  console.log("5) fetch(asJson)");
  const fetched = await client.fetch(pinned.url, { asJson: true });
  if (!fetched || typeof fetched !== "object" || fetched.marker !== marker) {
    throw new Error("fetched payload does not match marker");
  }

  console.log("6) delete");
  await client.delete(fileName);

  console.log("Integration flow passed", {
    operationId: stored.operationId,
    deterministicUrl: pinned.url,
    contentHash: pinned.contentHash,
  });
}

main().catch((err) => {
  console.error("Integration flow failed:", err);
  process.exit(1);
});

File: rust/tests/integration.rs
mod common;

use pipe_agent_storage::{OperationState, StoreOptions};
use serde_json::json;
use std::env;

#[tokio::test]
async fn integration_flow_prod_host() -> Result<(), Box<dyn std::error::Error>> {
    if env::var("PIPE_RUN_INTEGRATION_TESTS").ok().as_deref() != Some("1") {
        eprintln!("Skipping live integration test. Set PIPE_RUN_INTEGRATION_TESTS=1 to enable.");
        return Ok(());
    }

    let (mut client, needs_logout) = common::authenticated_client().await?;

    let file_name = format!("agent/integration-rust-{}.json", common::now_ms());
    let marker = format!("marker-{}", common::now_ms());

    println!("1) store(wait=false) file_name={}", file_name);
    let stored = client
        .store_json(
            &json!({ "marker": marker, "kind": "integration" }),
            StoreOptions {
                file_name: Some(file_name.clone()),
                wait: false,
                ..Default::default()
            },
        )
        .await?;

    let operation_id = stored
        .operation_id
        .ok_or("store did not return operation_id")?;

    println!("2) check_status operation_id={}", operation_id);
    let status = client.check_status(Some(&operation_id), None).await?;
    println!("   status={:?}", status.status);

    println!("3) wait_for_operation");
    let completed = client.wait_for_operation(&operation_id, None).await?;
    assert_eq!(completed.status, OperationState::Completed);
    println!(
        "   completed: content_hash={:?}, deterministic_url={:?}",
        completed.content_hash, completed.deterministic_url
    );

    println!("4) pin");
    // Use the deterministic_url from the completed status directly,
    // rather than re-querying by operation_id (avoids server-side race).
    let deterministic_url = completed
        .deterministic_url
        .ok_or("completed status missing deterministic_url")?;
    let pinned = client.pin(&deterministic_url).await?;
    assert!(!pinned.url.is_empty());

    println!("5) fetch_json");
    let fetched: serde_json::Value = client.fetch_json(&pinned.url).await?;
    assert_eq!(
        fetched.get("marker").and_then(|v| v.as_str()),
        Some(marker.as_str())
    );

    println!("6) delete");
    client.delete_file_name(&file_name).await?;

    if needs_logout {
        println!("7) logout");
        client.auth_logout().await?;
    }

    println!(
        "Integration flow passed: operation_id={}, deterministic_url={}",
        operation_id, pinned.url
    );

    Ok(())
}
